### ISSUE TITLE
Cross-compile for Dotty

### ISSUE THREAD

**Issue by jdegoes** [zio/zio-config#182]

**Comment by afsalthaj** [comment]
@jdegoes unfortunately magnolia doesn’t support 2.11. We will miss auto configuration when it is 2.11

**Comment by jdegoes** [comment]
@afsalthaj That's ok, auto-derivation is nice but not essential. The ordinary API is pretty minimal as it is.

**Comment by DamianReeves** [comment]
So what needs to happen to get this working? I'd love to use `zio-config` in a Spark application which is stuck on Scala 2.11. Willing to create a PR with some guidance.

**Comment by afsalthaj** [comment]
Yes, we do understand this scenario. However, current builds in zio projects couldn't be customised per scala version.

I will take this up with higher priority

**Comment by afsalthaj** [comment]
@Dmitrii-Shevchenko  I can see a few commits referenced to this issue.  That's cool.
Are you looking at solving this one ?

**Comment by afsalthaj** [comment]
@Dmitrii-Shevchenko  Let me know if you need any help. Could I assign this issue to you?

**Comment by afsalthaj** [comment]
@Dmitrii-Shevchenko Are you working on this issue? We need this to be cross-compiled with dotty soon :)

**Comment by afsalthaj** [comment]
This is in progress

**Comment by vigoo** [comment]
A few days ago when ZIO `1.0.5` was released I updated a couple of libraries and also gave a try to zio-config. For the magnolia based libs I guess we just wait for a Scala 3 magnolia, with the rest the only issue I founds it that the `ConfigDescriptor.apply`:

```scala
def apply[B](app: A => B, unapp: B => Option[A]): ConfigDescriptor[B]
```
is completely unusable in practice with Scala 3, because of the change in the generated `unapply` for products (https://scalacenter.github.io/scala-3-migration-guide/docs/incompatibilities/other-changed-features.html#explicit-call-to-unapply)

So although the library itself compiles fine, the tests and the user code using this in the `(Prod.apply, Prod.unapply)` style are not.

I have a proposal for it with a proof of concept implementation (simplified example outside of zio-config).

The proposal:
- deprecate the existing `apply` in favor of a new method `to[Product]`
- define a new method on config descriptor `def to[B](implicit conv: TupleConversion[B, A])`
- `TupleConversion` itself would be implemented separately for Scala 2 and Scala 3, both providing an implicit for any product type

Scala 3 version is quite simple:
```scala
trait TupleConversion[A, B] {
  def to(a: A): B
  def from(b: B): A
}

object TupleConversion {
  inline given auto[Prod <: Product](using m: Mirror.ProductOf[Prod]): TupleConversion[Prod, m.MirroredElemTypes] = 
    new TupleConversion[Prod, m.MirroredElemTypes] {
      def to(a: Prod): m.MirroredElemTypes = Tuple.fromProductTyped(a)
      def from(b: m.MirroredElemTypes): Prod = m.fromProduct(b)
    }
}
```

The equivalent Scala 2 version is a bit longer macro (~50 lines) that I'm not copying here but does the same. 

Let me know what you think, and if we want to do this I can provide my proof of concept macros and integrate them.
